src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala

```diff
@@ -93,6 +94,204 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  val A_P1 = "part=p1/a"
+  val B_P1 = "part=p1/b"
+  val C_P2 = "part=p2/c"
+  val D_P2 = "part=p2/d"
+  val E_P3 = "part=p3/e"
+  val F_P3 = "part=p3/f"
+  private val addA_P1 = AddFile(A_P1, Map("part" -> "p1"), 1, 1, dataChange = true)
+  private val addB_P1 = AddFile(B_P1, Map("part" -> "p1"), 1, 1, dataChange = true)
+  private val addC_P2 = AddFile(C_P2, Map("part" -> "p2"), 1, 1, dataChange = true)
+  private val addD_P2 = AddFile(D_P2, Map("part" -> "p2"), 1, 1, dataChange = true)
+  private val addE_P3 = AddFile(E_P3, Map("part" -> "p3"), 1, 1, dataChange = true)
+  private val addF_P3 = AddFile(F_P3, Map("part" -> "p3"), 1, 1, dataChange = true)
+
+  test("block concurrent commit when table is not configured") {
+    withLog(addA_P1 :: addD_P2 :: addD_P2.remove :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 reads A_P1
+      tx1.filterFiles()
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies P2
+      tx2.commit(addC_P2 :: addD_P2.remove :: Nil, Truncate())
+
+      intercept[ConcurrentWriteException] {
+        // this is safe however table was not configured for concurrent writes
+        tx1.commit(addE_P3 :: addF_P3 :: Nil, Truncate())
+      }
+    }
+  }
+
+  test("allow concurrent commit independent partitions") {
+    withConcurrentLog(addA_P1 :: addD_P2 :: addD_P2.remove :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 reads A_P1
+      tx1.filterFiles()
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies only P2
+      tx2.commit(addC_P2 :: addD_P2.remove :: Nil, Truncate())
+
+      // TX1 free to commit P3
+      tx1.commit(addE_P3 :: addF_P3 :: Nil, Truncate())
+      checkAnswer(log.update().allFiles.select("path"),
+        // start
+        Row(A_P1) ::
+        // TX2
+        Row(C_P2) ::
+        // TX1
+        Row(E_P3) :: Row(F_P3) :: Nil)
+    }
+  }
+
+  test("allow concurrent commit when read from other partition") {
```

Comment:
it would also be nice to have an analogous test when tx1 was reading p1, to verify that that gets blocked.

Reply:
added test `"block concurrent commit when read specific conflicting partition"`